File: Jump_to_Algorithm3/week14/SeoyeonHong/1238_파티.py
# N명의 학생들 중 오고 가는데 가장 오래 걸리는 학생의 소요시간
import sys
import heapq

input = sys.stdin.readline
N, M, X = map(int, input().split())
INF = int(1e9)
load = [[] for _ in range(N+1)]
load_reversed = [[] for _ in range(N+1)]
max_time = 0

def dijkstra(l, X): # X에서 각 정점까지 가는데 걸리는 최소시간
    min_time = [INF for _ in range(N+1)] # min_time[i]: i번째 정점까지 가는데 걸리는 최소시간
    min_time[X] = 0
    q = [(0, X)]
    heapq.heapify(q)
    while q:
        total_time, A = heapq.heappop(q)
        for B, time in l[A]:
            if min_time[B] > total_time + time:
                min_time[B] = total_time + time
                heapq.heappush(q, (min_time[B], B))
    return min_time
                

for _ in range(M):
    s, e, t = map(int, input().split())
    load[s].append((e, t))
    load_reversed[e].append((s, t))

from_X = dijkstra(load, X) # X에서 모든 정점까지의 최단 시간, 정방향 그래프 + 다익스트라
to_X = dijkstra(load_reversed, X) # 모든 정점에서 X까지의 최단 시간, 역방향 그래프 + 다익스트라
max_time = max(to_X[i] + from_X[i] for i in range(1, N+1))
    
print(max_time)

File: Jump_to_Algorithm3/week14/SoohyunKim/1238_파티.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

function solution(input) {
  const [N, M, X] = input[0].split(' ').map(Number);
  const roads = input.slice(1).map((line) => line.split(" ").map(Number));

  const graph = Array.from({ length: N + 1 }, () => []);
  const reverseGraph = Array.from({ length: N + 1 }, () => []);

  for (const [from, to, time] of roads) {
    graph[from].push([to, time]);
    reverseGraph[to].push([from, time]);
  }

  const toX = dijkstra(X, reverseGraph, N);
  const fromX = dijkstra(X, graph, N);

  let maxTime = 0;

  for (let i = 1; i <= N; i++) {
    maxTime = Math.max(maxTime, toX[i] + fromX[i]);
  }

  return maxTime;
}

function dijkstra(start, graph, N) {
  const distances = Array(N + 1).fill(Infinity);
  const visited = Array(N + 1).fill(false);
  distances[start] = 0;

  const priorityQueue = [[0, start]];

  while (priorityQueue.length) {
    priorityQueue.sort((a, b) => b[0] - a[0]); 
    const [currentDist, currentNode] = priorityQueue.pop();

    if (visited[currentNode]) continue;
    visited[currentNode] = true;

    for (const [nextNode, weight] of graph[currentNode]) {
      const newDist = currentDist + weight;

      if (newDist < distances[nextNode]) {
        distances[nextNode] = newDist;
        priorityQueue.push([newDist, nextNode]);
      }
    }
  }

  return distances;
}

console.log(solution(input));